### PR TITLE
pass package_name explicitly in click.version_option decorators

### PR DESCRIPTION
**Pull request by nicoa**
Closes https://github.com/jazzband/pip-tools/issues/1398 .

Unfortunately mypy threw some errors during pre-commit run, where two of them seem to be valid (coming from CI as well).

Further, some of the tests failed, but this seems to be related to a pip.conf file adding extra indices to the compiled requirements, I'll re-run them again to see if this is related.

**Changelog-friendly one-liner**: pass package_name explicitly in click.version_option decorators for compatibility with click>=8.0

##### Contributor checklist

- [ ] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by nicoa** [comment]
the failing mypy CI is due to click not being upgraded yet in typeshed. What do you think, should I add a `# type: ignore` to the two respective lines or wait for typeshed to incorporate click 8 changes?

**Comment by nicoa** [comment]
Update: see https://github.com/python/typeshed/issues/5503, this seems to be an issue with click directly. same questions still applies: ignore that typing?

**Comment by davidism** [comment]
Click 8 provides typing directly, typeshed stubs are not used anymore. Not sure how the pre-commit mypy hook works, are you sure you're getting the latest click version in that environment?

**Comment by nicoa** [comment]
Hey, are any maintainers around here and reading this? It seems that the pre-commit CI is optional for merging, would be great to get some comment on whether it's preferred to first merge that or first fix the failing CI job.

Regarding the mypy hook (the failing one): it seems that `additional_dependencies: [click>=8.0.0]` in pre-commit-config fixes the wrong click version, but introduces a whole lot of new errors (probably due to switch to new click version, not sure). I can provide them here if required. Would it make sense to open a ticket to fix them?

**Comment by atugushev** [comment]
> Regarding the mypy hook (the failing one): it seems that `additional_dependencies: [click>=8.0.0]`

Aw, that makes sense now, since we cache installed dependencies in CI. @webknjaz Is there a decent way to drop the cache (other than update the cache key or example)?

**Comment by nicoa** [comment]
> > Regarding the mypy hook (the failing one): it seems that `additional_dependencies: [click>=8.0.0]`
> 
> Aw, that makes sense now, since we cache installed dependencies in CI. @webknjaz Is there a decent way to drop the cache (other than update the cache key or example)?

Still, there would be other errors introduced then: I'd open a separate ticket for that, would you agree with that procedure?

**Comment by atugushev** [comment]
> Still, there would be other errors introduced then with the new click version: I'd open a separate ticket for that, would you agree with that procedure?

Yeah, let's track this in a separate issue. Thanks! In addition to click, I'm inclined to think that the latest pip update brought new types and we need to address that also.

**Comment by webknjaz** [comment]
> Is there a decent way to drop the cache (other than update the cache key or example)?

Not that I know of. Although, you seem to be assuming the wrong check. The one that is failing is not coming from GHA. It comes from the pre-commit.ci service.

**Comment by nicoa** [comment]
The newest pre-commit ci run seemed to use the proper version and raised the same errors I experienced when explicitly adding the new dependency (compare #1401 and #1402 ). I'd say this can be merged while ignoring the failing pre-commit.ci check, what do you think @webknjaz ?

**Comment by webknjaz** [comment]
Yes, unless we want to have broad compatibility with older click, we should bump the dep too.

Regarding broken typing, if it's directly connected to this change, it must be fixed as a part of it.

**Comment by nicoa** [comment]
will bump the dependency then.

Regarding the typing: no, it's due to new click and is on latest pre-commit-CI as well, this just didn't had it due to some caching apparently. Would be solved in #1402 .

**Comment by webknjaz** [comment]
But bumping click is directly related to this then. This PR cannot be accepted with breaking typing.

**Comment by nicoa** [comment]
Sorry, that was a typo... I wanted to write that it's due to new pip version - so not related in any way to click and the current CI setup is broken with the same errors if running now as well already (as you can see in #1401 ).

**Comment by webknjaz** [comment]
Oh, I misunderstood that. Thinking more about dropping the older click, I think it's probably preferable to support those. This is because click is fairly popular and there's a good chance that some users will have other tools in the same env that may pin click more strictly and so it's a potentially serious conflict we'd cause.

File: piptools/scripts/compile.py
Comment (nicoa) [inline piptools/scripts/compile.py:48]: @webknjaz should we rather not still stick to click >= 7 and pass this option only on click version 8?

**Comment by nicoa** [comment]
fair point, but not sure how to handle as the current setup is broken sometimes under click 8. Made some inline suggestion in the code to keep click 7 but handle 8 separately then.

**Comment by webknjaz** [comment]
There's a few ways:
1)
```python
ver_kwargs = {} if click7 else {'package_name': 'piptools'} 
...
@click.version_option(**ver_kwargs)
def cli(...): ...
```
2)
```python
...
def cli(...): ...

if click8:
    cli = click.version_option(package_name='piptools')(cli)
```
3)
```python
from itertools import partial
ver_opt = click.version_option if click7 else partial(click.version_option, package_name='piptools')
...
@ver_opt()
def cli(...): ...
```

**Comment by nicoa** [comment]
That's what I suggested, sorry if that was not clear :-)  was just waiting for the okay to do this version-based differencing, would implement option 1 then. Thanks!

File: piptools/scripts/compile.py
Comment (webknjaz) [inline piptools/scripts/compile.py:37]: This check is not very reliable: what if the version is 70?

Comment (nicoa) [reply piptools/scripts/compile.py:37]: true, I wrongly expected this to be removed sooner, but better to be safe. What do you think about `click.__version__.split(".")[0] == "7"`? That looks uglier but would do what it should do.

Comment (webknjaz) [reply piptools/scripts/compile.py:37]: Almost. Except you should turn that into an integer and use `>` in the comparison. You may also want to move this into a compat module and declare some variable like `IS_CLICK_VER_8_PLUS`.

The reason for sooner vs later is not the best way of thinking about it is because over time the end users may use a release of pip-tools with this code which allows for any version of click above 7. So they may end up with old pinned pip-tools but a very modern click. I seem to recall stories of how win devs used a comparison with "9" for win 95/98 and then suddenly it was 2000, and then win 7/8 which started driving nuts old software that still works on new platforms because people never considered that versioning could change, back in the day.

Comment (nicoa) [reply piptools/scripts/compile.py]: Definitely makes sense, thanks for the words!

implemented this, hope it is the proper place and implementation :)

File: piptools/_compat/__init__.py
Comment (webknjaz) [inline piptools/_compat/__init__.py:4]: Side note: why is this a list? It's usually recommended to make `__all__` immutable, hence use a tuple.

P.S. This is out of the scope of this PR, just wanted to make a note.

File: piptools/_compat/click_compat.py
Comment (webknjaz) [inline piptools/_compat/click_compat.py:4]: I don't think the var name corresponds to its contents. Either add that it's major or only store the version.

File: piptools/_compat/click_compat.py
Comment (webknjaz) [inline piptools/_compat/click_compat.py:2]: I don't really like relying on how and whether pip vendors its deps. Plus it may cause problems when repackaged in distros that unbundle vendored software.

File: piptools/_compat/__init__.py
Comment (nicoa) [reply piptools/_compat/__init__.py:4]: to be fair, this is a list as well in the python tutorial on modules: https://docs.python.org/3/tutorial/modules.html#importing-from-a-package and the pip package has this format as well. 
Do you want to change this here or should I leave as it is? I actually don't have any preference here.

File: piptools/_compat/click_compat.py
Comment (nicoa) [reply piptools/_compat/click_compat.py:1]: thought it makes sense to have this analog to `pip_compat.py`, but can change this if you prefer `click` over that

File: piptools/_compat/click_compat.py
Comment (nicoa) [reply piptools/_compat/click_compat.py:2]: sure, was just in analogy to `pip_compat` module - would you agree with using `int(click.__version__.split(".")[0])`?

File: piptools/_compat/__init__.py
Comment (webknjaz) [reply piptools/_compat/__init__.py:4]: No, don't touch this because it's out of the scope. But in the future, this should probably be added. I've added this comment for history.

P.S. Normally, pylint would warn about this problem but I don't think it's integrated.

File: piptools/_compat/click_compat.py
Comment (webknjaz) [reply piptools/_compat/click_compat.py:2]: Yeah, that should be enough.

Comment (webknjaz) [reply piptools/_compat/click_compat.py:2]: Although, you may want to multi-line it to keep the inline complexity low

**Review by webknjaz — APPROVED** [review]

File: piptools/_compat/click_compat.py
Comment (nicoa) [reply piptools/_compat/click_compat.py:2]: not sure what you mean with multi-lining, as for me black would merge this into one line if not forcing to not do on the following: 
```
int(
    click.__version__.split(".")[0]
)
```

File: piptools/_compat/__init__.py
Comment (nicoa) [reply piptools/_compat/__init__.py:4]: at least pycodestyle seems to do so, compare https://github.com/PyCQA/pydocstyle/issues/62

File: piptools/_compat/click_compat.py
Comment (nicoa) [reply piptools/_compat/click_compat.py:1]: fair point, would adapt then 👍

File: piptools/_compat/click_compat.py
Comment (nicoa) [reply piptools/_compat/click_compat.py:2]: nvmd, will simply add a comment then in between the brackeets

**Comment by nicoa** [comment]
Hey @webknjaz thanks for your patience with me during this PR! For me it looks like it is ready to get into a next release. Whats the progress from here? would you merge it to master directly and/or add a milestone (I am not able to do any of those two)? And do I have to do anything from my side or does everything look good now?

File: piptools/_compat/click_compat.py
Comment (webknjaz) [reply piptools/_compat/click_compat.py]: Yeah, that's one of the reasons I dislike black...

As for the line complexity, I like this metric that https://wemake-python-stylegui.de introduced called Jones complexity. It really makes a difference maintainability/readability-wise!

Here's some more info on this https://sobolevn.me/2019/10/complexity-waterfall#lines.

I wish this set of linter plugins was integrated here.

**Comment by webknjaz** [comment]
So it looks like the CI is failing because of an unrelated problem so that would need to be fixed before merging this PR.

Regarding the release, I don't do those so you'll have to wait.

File: piptools/_compat/click_compat.py
Comment (nicoa) [reply piptools/_compat/click_compat.py]: didn't knew the blog entry regarding the complexity until now, found it a good read and would recommend to colleagues, thanks a lot for that, very interesting!

**Comment by nicoa** [comment]
Alright, would wait then, thanks for the information! Yes, for the CI there is issue #1403 .

**Comment by nicoa** [comment]
Hey @webknjaz , as you already reviewed this I'd merge it as well, but not sure if this is only fine for more senior persons on the project. Should I proceed or do you / others want to keep the hands on what gets to master?

**Comment by webknjaz** [comment]
@nicoa feel free to merge

**Pull request merged**
